### ISSUE TITLE
Parsing '2016-22-99' returns valid date

### ISSUE THREAD

**Issue by andrewn** [date-fns/date-fns#347]
Parsing a date with an invalid month and year such as:
  `dateFns.parse('2016-22-99')`

Returns:
  `"2018-01-06T23:00:00.000Z"`

Doing the same in Chrome 54.0, Safari 10 and node 7.1.0 returns an `Invalid Date`.

I'm not sure who's right but I'd expect this to be invalid?

**Comment by LeshaKoss** [comment]
@andrewn 
> Doing the same in Chrome 54.0, Safari 10 and node 7.1.0 returns an Invalid Date.

Do you mean like `new Date('2016-22-99')`?

Well, 99th of 22nd month 2016 is a valid but non-existent date, in the sense that `new Date(2016, 21, 99)` would yield the same result. ATM in `parse` we don't check if the date exist, and if we will, it would make a lot of problems to solve. For example, some dates exist in some timezones but don't in the other timezones:
```
$ env TZ="America/New_York" node
> new Date(2017, 2 /* March */, 12, 2).toString()
'Sun Mar 12 2017 03:00:00 GMT-0400 (EDT)'
> new Date(2017, 2 /* March */, 12, 3).toString()
'Sun Mar 12 2017 03:00:00 GMT-0400 (EDT)'
```

In `America/New_York` timezone the date `2017-03-12T02:00:00` don't exist because DST. Should we return `Invalid Date` in this case too to make it consistent? Well, no, that would be certainly an unexpected result for users. But what about 31st of February? `Date` constructor creates a valid Date despite it can never exist:
```
> new Date('2016-02-31').toString()
'Tue Mar 01 2016 19:00:00 GMT-0500 (EST)'
```
Also `new Date('2016-02-01T24:00:00')` creates a valid date too.

I don't see the reason to return an `Invalid Date` for 32nd days of months but not for 31st days of short months. We chose an easiest path and don't check existence of dates in `parse` and just overflow extra days to the next month, extra months to the next year etc. Please tell me if you consider this behaviour to be confusing and if we should change it in the future

**Comment by halt-hammerzeit** [comment]
@LeshaKoss Then it should be added to `date-fns` cons section:

> date-fns cons is:
Just 7 23 locales supported (help is welcome!).
No alternative to Moment Timezone (yet).
Functionality-wise date-fns equals to Moment.js.

They're obviously not equal to Moment.js: moment.js can parse and validate inputted dates while date-fns can't.

**Comment by andrewn** [comment]
@LeshaKoss Thanks for the detailed response, and sorry for not replying sooner.

> Do you mean like new Date('2016-22-99')?

Yes, I  was doing `new Date('2016-22-99')` on all the platforms I mentioned and they all return `Invalid Date`.

First of all, I'm only considering ISO calendar date strings and not times or timezones as that's what my application is consuming, so my issue is only concerning parsing date strings of the form "YYYY-MM-DD".

The date parsing logic for Chrome seems to be that as long as the ISO month and date components are within the allowable range then it will overflow the date but if it's outside the range then it returns an `Invalid Date`.

e.g. `(new Date('2016-02-31')).toISOString()` returns `2016-03-02T00:00:00.000Z` because `31` is within the range 1-31 for months.

`(new Date('2016-02-32')).toISOString()` returns `Invalid Date` because `32` is outside of `1-31`. 

The same seems to happen with months where any month in the range `1-12` is valid but anything outside of that returns `Invalid Date`.

I take your point that technically, if `2016-02-31` is overflowed then so should `2016-02-32`. However, I'd argue as a human user of the library, I'd expect something like `2016-22-99` to be "invalid" as an input date and not overflow to a date far into the future. That behaviour is very confusing to me. I think if `2016-02-31` overflowed however, then I'd be a little surprised but not too surprised. ;-)

I totally understand if that's too complex to support but I thought I'd raise it here in case it was a bug.

I really love date-fns by the way, thanks so much for your work on it!

**Comment by logaretm** [comment]
Hello, Is this still being debated?

First of all thanks for the great work. I really liked the compactness of this library and have a lot of useful utilities, much appreciated.

I recently decided to use `date-fns` as the date engine for my library `vee-validate` (validation plugin for Vue.js framework) instead of moment, since moment currently lacks such modularity and I don't really need most of its functionality, plus having the users to include/install moment is not very ideal since I wanted to include the date validations without dependencies.

I understand your design choice and appreciate sticking with the native js behavior, but that makes it harder for validation authors to make use of `date-fns` in some basic operations like parsing and comparison (after, before, between, etc ...) like @halt-hammerzeit pointed out.

Now changing the current behavior might be breaking or impractical at the moment, but having a way to detect such overflows would be more than welcome, maybe adding an option to the parse method like this:

```js
// valid date, overflows
let date = parse('YYYY-MM-DD', '2016-22-99', new Date());

// invalid date
date = parse('YYYY-MM-DD', '2016-22-99', new Date(), { exists: true });
```
or an entirely new `parseExists` method whichever you think is best, again; many thanks for your awesome work 💯

**Comment by ThomHurks** [comment]
@leshakoss I'm a ```vee-validate``` user but also use ```date-fns``` separately in my project. I didn't even know that ```parse``` allows overflowing like this until I read the report. Useful to know!

Both for usage in ```vee-validate``` and for my own use I would love a function/option that imposes more restrictions when parsing. I understand that figuring out when a date "exists" or not can be tricky, especially when you take timezones into account, but I assume that moment.js doesn't do that as well. I think we can all agree that there will never be a 22nd month in a year or a 99th day in a month, so those dates would be invalid. The edge cases that you describe would simply pass this kind of check.

So simply put: make it optional, check for all the obvious and straightforward cases, and if people want to do more advanced checks with specific timezones etc, then they are free to write that code themselves ;) Does that approach make sense?

**Comment by robport** [comment]
It seems like a slam dunk of a use case to know if your user has entered a date that exists vs one that doesn't.  I don't want to allow users to enter non-existent payment dates, for example.

**Comment by leshakoss** [comment]
Fixed in [v2.0.0-alpha.14](https://gist.github.com/kossnocorp/a307a464760b405bb78ef5020a4ab136#v200-alpha14) in PR  #808
